rclcpp/src/rclcpp/time_source.cpp

```diff
@@ -141,6 +149,10 @@ void TimeSource::detachNode()
   sim_time_cb_handler_.reset();
   node_parameters_.reset();
   disable_ros_time();
+  if (clock_executor_thread_.joinable()) {
+    clock_executor_.cancel();
+    clock_executor_thread_.join();
+  }
```

Comment:
Looks like if the destructor runs before before the `clock_executor_thread` manages to call `spin`, this will block for ever.
Seems unlikely, but there are several bug reports for this behaviour already in other areas.
Perhaps set a flag after `spin`, and keep calling `cancel` until the flag is true?

Reply:
which is https://github.com/ros2/rclcpp/issues/1454, you are right. there's been discussion on that, it probably would take some time to support that. in the meantime, we'd want to have atomic internal flag which indicate spinning or not?

Reply:
I have tested to attach and detach several times in a raw a Node to a TimeSource, checking if the thread is running in between:
```c
auto node = std::make_shared<rclcpp::Node>("test_node");
node->set_parameter(rclcpp::Parameter("use_sim_time", true));

for(int i = 0; i < 10; i++) {
  ts.attachNode(node);
  std::cout << "[Node attached] Clock thread is running? : "  <<
               ts.IsClockThreadJoinable() << std::endl;
  ts.detachNode();
  std::cout << "[Node detached] Clock thread is running? : "  <<
               ts.IsClockThreadJoinable() << std::endl;
  std::cout << std::endl;
}
```
I didn't notice any unexpected behavior: the thread is running when the node is attached, and is not otherwise, and nothing is blocking.

Reply:
it is unlikely to happen because of context switches and race condition. but in theory that happens, and if that happens it will hang up eventually. as far as i see, this needs to be taken care by atomic flag or something, at least https://github.com/ros2/rclcpp/issues/1454 is addressed.

Reply:
Ok, it is done [here](https://github.com/ros2/rclcpp/pull/1556/commits/aa2e216e2043f5e23ea2c366cc5bae9deb6bf816).

Reply:
because `is_clock_executor_cancelled_` is accessed by different threads, I think you'll need to make it `atomic<bool>`
Also, maybe you want a small sleep or yield that destructor, after `cancel` to allow the other thread to run and set the boolean.

And just for reference, there is another approach to dealing with this issue described here https://github.com/ros2/rclcpp/issues/1454#issuecomment-799695171

Reply:
> I think you'll need to make it atomic<bool>

Just FYI, it is something like https://github.com/ros2/rclcpp/blob/35c89c8afcac89a43872f0bf5b763ff324c82b05/rclcpp/include/rclcpp/executor.hpp#L524